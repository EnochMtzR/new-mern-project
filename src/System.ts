import { Services as _Services } from "./Services/Services";
import { LocalStore as _Store } from "./store/Store";
import * as locale from "./locales/data.json";
import * as moment from "moment";

/**Holds all InHotel's code necesary to talk to **_Services_** or to affect the **_Store_**
 * As well as some global **_Interfaces_** and **_functions_**.
 */
export namespace System {
  export import Services = _Services;
  export import Store = _Store;

  export const getLocaleText = (id: string, language: string) => {
    const localeData: locale.LocaleData = locale;
    return localeData[language][id];
  };

  export enum ReservationStatus {
    NOT_GUARANTEED = "1",
    GUARANTEED = "2",
    IN_HOUSE = "4",
    DEPARTURE = "5"
  }

  export interface Room {
    id: string;
    type: string;
  }

  export interface Availability {
    date: moment.Moment;
    availability: number;
  }

  export class RoomBlock {
    id: string;
    startDate: moment.Moment;
    endDate: moment.Moment;
    roomId: string;
  }

  export class Reservation extends RoomBlock {
    status: string;
    hasASignedRoom: boolean;
    guest: string;
    agency: string;
    notes: string;

    set arrivalDate(value: moment.Moment) {
      this.startDate = value;
    }

    get arrivalDate() {
      return this.startDate;
    }

    set departureDate(value: moment.Moment) {
      this.endDate = value;
    }

    get departureDate() {
      return this.endDate;
    }
  }

  export class OutOfService extends RoomBlock {
    description: string;
  }

  export interface RoomAvailabilityResponse {
    roomTypes: string[];
    rooms: Room[];
    availability: Availability[];
    reservations: Reservation[];
    OutOfService: OutOfService[];
  }

  export const cleanTooltipAndContextMenu = () => {
    let state = JSON.parse(localStorage.getItem("state")) as System.Store.State;
    if (state) {
      state.config.contextMenu = undefined;
      state.config.toolTip = "";

      localStorage.setItem("state", JSON.stringify(state));
    }
  };
}

declare global {
  interface String {
    /** Capitalizes the first letters of each word, except conjunctions and prepositions */
    toTitleCase(): string;
  }

  interface Math {
    /**
     * Calculates the **percentage** based on the **value** and **total** given.
     *
     * **_Note: the base factor (which defaults to 100) is the whole that the total represents._**
     *
     * ex. `total=5 represents the 100%; value=1 represents ?%`
     *
     * `total=157 represents 360째 value=2 represents ?째`
     *
     */
    percentageFromValue(
      value: number,
      total: number,
      baseFactor?: number
    ): number;

    /**
     * Calculates the **percentage** based on the **value** and **total** given.
     *
     * **_Note: the base factor (which defaults to 100) is the whole that the total represents._**
     *
     * ex. `total=5 represents the 100%; percentage=10% represents ?`
     *
     * `total=157 represents 360째 value=90째 represents ?`
     *
     */
    valueFromPercentage(
      percentage: number,
      total: number,
      baseFactor?: number
    ): number;
  }
}

Math.percentageFromValue = function(
  value: number,
  total: number,
  baseFactor = 100
) {
  return (value * baseFactor) / total;
};

Math.valueFromPercentage = function(
  percentage: number,
  total: number,
  baseFactor = 100
) {
  return (percentage * total) / baseFactor;
};

String.prototype.toTitleCase = function() {
  var i, j, str, lowers, uppers;
  str = this.replace(/([^\W_]+[^\s-]*) */g, function(txt: string) {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
  });

  // Certain minor words should be left lowercase unless
  // they are the first or last words in the string
  lowers = [
    "A",
    "An",
    "The",
    "And",
    "But",
    "Or",
    "For",
    "Nor",
    "As",
    "At",
    "By",
    "For",
    "From",
    "In",
    "Into",
    "Near",
    "Of",
    "On",
    "Onto",
    "To",
    "With"
  ];
  for (i = 0, j = lowers.length; i < j; i++)
    str = str.replace(new RegExp("\\s" + lowers[i] + "\\s", "g"), function(
      txt: string
    ) {
      return txt.toLowerCase();
    });

  // Certain words such as initialisms or acronyms should be left uppercase
  uppers = ["Id", "Tv"];
  for (i = 0, j = uppers.length; i < j; i++)
    str = str.replace(
      new RegExp("\\b" + uppers[i] + "\\b", "g"),
      uppers[i].toUpperCase()
    );

  return str;
};

document.oncontextmenu = () => false;
