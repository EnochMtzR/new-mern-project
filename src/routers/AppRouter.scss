.router {
  height: 100%;
  display: grid;
  grid-template-rows: 62.19px auto;
}

@media screen and (min-width: 660px) {
  .router {
    grid-template-rows: 56.29px auto;
  }
}
