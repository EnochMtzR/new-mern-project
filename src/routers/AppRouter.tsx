import * as React from "react";
import { BrowserRouter, Route, Switch } from "react-router-dom";
import { IntlProvider } from "react-intl";
import { connect } from "react-redux";
import { System } from "../System";

//Components import
import { ToolTip } from "../ToolTip/ToolTip.component";
import { ContextMenu } from "../ContextMenu/ContextMenu.component";
import { NotFound } from "../not-found-page/NotFound.page";
import * as locale from "../locales/data.json";
import { Dispatch } from "redux";

interface StateProps {
  user: System.Store.User.State;
  config: System.Store.Config.State;
}

interface ComponentState {
  isHeaderVisible: boolean;
  isHeaderSticky: boolean;
  wasHeaderUnsticked: boolean;
  localeData: locale.LocaleData;
}

const defaultState: ComponentState = {
  isHeaderVisible: false,
  isHeaderSticky: false,
  wasHeaderUnsticked: false,
  localeData: locale
};

require("./AppRouter.scss");

export class AppRouter extends React.Component<StateProps, ComponentState> {
  state = defaultState;

  constructor(props: StateProps) {
    super(props);
  }

  render() {
    return (
      <IntlProvider
        locale={this.props.config.language}
        messages={this.state.localeData[this.props.config.language]}
        key={this.props.config.language}
      >
        <BrowserRouter>
          <div className={this.props.user.id ? "router" : ""}>
            <Switch>
              <Route component={NotFound} />
            </Switch>
          </div>
        </BrowserRouter>
      </IntlProvider>
    );
  }
}

export const mapStateToProps = (state: System.Store.State) => ({
  user: state.user,
  config: state.config
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AppRouter);
