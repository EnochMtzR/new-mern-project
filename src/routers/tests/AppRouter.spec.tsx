import * as React from "react";
import { shallow } from "enzyme";
import { AppRouter, mapStateToProps } from "../AppRouter";
import { System } from "app/System";

const { User, Config, defaultState } = System.Store;

describe("Testing AppRouter.tsx", () => {
  test("Should render component Correctly", () => {
    const wrapper = shallow(
      <AppRouter user={User.defaultState} config={Config.defaultState} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("Should return correct state in mapStateToProps()", () => {
    const returnedState = mapStateToProps(defaultState);
    expect(returnedState).toEqual(defaultState);
  });
});
