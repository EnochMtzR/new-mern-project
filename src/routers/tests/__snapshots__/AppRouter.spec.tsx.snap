// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AppRouter.tsx Should render component Correctly 1`] = `
<IntlProvider
  key="es"
  locale="es"
  messages={
    Object {
      "Footer.language.english": "Inglés",
      "Footer.language.spanish": "Español",
      "Header.about": "Acerca de Nosotros",
      "Header.modules": "Módulos",
      "Header.news": "Noticias",
      "Header.support": "Soporte",
      "Header.userMenu.signIn": "Iniciar Sesión",
      "LoginForm.errorDisplay": "¡Oops! Parece que las credenciales provistas son incorrectas. Por favor intente denuevo.",
      "LoginForm.login": "Entrar",
      "LoginForm.password": "Contraseña",
      "LoginForm.user": "Usuario",
    }
  }
>
  <BrowserRouter>
    <div>
      <Header
        invisible={true}
      />
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/signIn/"
        />
        <Route
          component={[Function]}
        />
      </Switch>
      <Connect(Footer) />
    </div>
  </BrowserRouter>
</IntlProvider>
`;
